posts/2022/02/making-gpus-accessible-to-pydata-ecosystem-via-array-api.md

```diff
@@ -0,0 +1,484 @@
+  Python, critical to performance parts are written using compiled code extensions. These parts
+  are dependent on NumPy's internal memory representation, which makes them CPU only. This
+  restricts the usage Array API to dispatch such code. There is another project `uarray`
+  for dispatching compiled code, which needs to be used along with Array API dispatching to make
+  it fully generic.
+
+See for instance following piece of code:
+
+```python
+from scipy import sparse
+
+graph = sparse.coo_matrix(<input_arrays>)
+```
+
+Here the `sparse.coo_matrix` is a compiled code extension in the `scipy.sparse` module. We need the
+dispatch to `cupyx.scipy.sparse.coo_matrix` for CuPy arrays and `scipy.sparse.coo_matrix` for NumPy
+arrays. This is not possible with the Array API's dispatching via `get_namespace`. For this demo we have
+added a workaround, until we have implemented the support for `uarray` based backed in SciPy. The
+workaround looks something like this for CuPy:
+
+```python
+xp, _ = get_namespace(<input_arrays>)
+
+if xp.__name__ == 'cupy.array_api':
+  from cupyx.scipy import sparse as cupy_sparse
+  coo_matrix =  cupy_sparse.coo_matrix
+else:
+  coo_matrix = scipy.sparse.coo_matrix
+
+graph = coo_matrix(<input_arrays>)
+```
+
+Note that, this workaround is only valid for NumPy/CuPy arrays.
+
+
+#### 2. Not implemented functions
+
+There are still a bunch of functions in the NumPy API which don't
+have a direct equivalent in the Array API, either due to the fact that those functions are not
+generic enough to be defined in the Array API or we haven't got around implementing those.
+
+Some examples of unimplemented functions that we encountered during this excercise:
+
+
+* `np.atleast_3d`
+* `np.strides`
+* `np.real`
+* `np.clip`
+
+#### 3. Inconsistencies with the NumPy API and Array API
+
+Although the Array API has managed to
+  get good coverage of the NumPy API, there are places where there are inconsistencies between
+  the two, it could be either with the function parameters or behaviour or the function or name
+  of the function like `concat` and `concatenate`.
+
+* Naming inconsistency between NumPy API and Array API
+
+```ipython
+
+In [1]: import numpy as np
+In [2]: import numpy.array_api as npx
+
+# Numpy API
+In [3]: np.concatenate
+Out[3]: <function numpy.concatenate>
+
+# Array API
+In [4]: npx.concat
+Out[4]: <function numpy.array_api._manipulation_functions.concat...>
+```
+
+* Indexing inconsistency between NumPy API and Array API
+
+```ipython
+In [1]: import numpy as np
+In [2]: import numpy.array_api as npx
+
+# Works with NumPy API
+In [3]: x = np.arange(10)
+In [4]: x[np.arange(2)]
+Out[4]: array([0, 1])
+
+
+# Doesn't Works with Array API
+In [5]: z = npx.arange(10)
+In [6]: z[npx.arange(2)]
+-----------------------------------------------------------------
+...
+IndexError: Non-zero dimensional integer array indices are not allowed in the array API namespace
+```
+
```

Comment:
This challenge section should probably conclude with something like: _"Each of the inconsistencies can be worked around. Renames may require some detective work but are typically easy code changes. Differences in behavior can be harder to deal with. The Array API standard aims to provide a set of functionality that's orthogonal but complete - so if it's possible to perform some array operation on accelerator devices, then there's typically a way to access it in the standard."_